### ISSUE TITLE
$null filter in relation not working

### ISSUE THREAD

**Issue by daniel-maegerli** [ppetzold/nestjs-paginate#538]
Hi, I've been fiddling around with an issue for a while now, which looks like a bug to me. I want to filter on a relation's property "viewedAt" (which is datetime or NULL in the database) but cannot make it _entirely_ work. Configuration (excerpt):
```
const paginatedProfiles = await paginate(query, profiles, {
  relations: ['matches'],
  filterableColumns: {
    'matches.viewedAt': [FilterOperator.NULL],
  },
});
```
When I do a request with `&filter.matches.viewedAt=$null` everything gets delivered, no matter if the "viewedAt" is NULL or not. 
When I do a request with `&filter.matches.viewedAt=$not:$null` it actually delivers to correct items, having "viewedAt" set (although I haven't even set the `FilterSuffix.NOT` in the configuration).

Payload 1 (excerpt):
```
{
  "meta": {
    "itemsPerPage": 10,
    "totalItems": 2,
    "currentPage": 1,
    "totalPages": 1,
    "filter": {
      "matches.viewedAt": "$null"
    }
  },
  "links": {
    "current": "http://localhost:8400/profiles/9f5fdc67-6541-45b1-ae61-8f6b25627a11?page=1&limit=10&filter.matches.viewedAt=$null"
  },
  "data": [
    {
      "match": {
        "matchingScore": 48.3,
        "viewedAt": "2023-03-15T09:15:34.000Z"
      }
    },
    {
      "match": {
        "matchingScore": 34.2,
        "viewedAt": null
      }
    }
  ]
}
```

Payload 2 (excerpt):
```
{
  "meta": {
    "itemsPerPage": 10,
    "totalItems": 1,
    "currentPage": 1,
    "totalPages": 1,
    "filter": {
      "matches.viewedAt": "$not:$null"
    }
  },
  "links": {
    "current": "http://localhost:8400/profiles/9f5fdc67-6541-45b1-ae61-8f6b25627a11?page=1&limit=10&filter.matches.viewedAt=$not:$null"
  },
  "data": [
    {
      "match": {
        "matchingScore": 48.3,
        "viewedAt": "2023-03-15T09:15:34.000Z"
      }
    }
  ]
}
```

Any idea? I've tried different versions and am on latest release (5.0.4) now. Still no luck to make it work.

**Comment by ppetzold** [comment]
Hmm, unfortunately we don't have coverage for `$null` in a one-to-many relation. Only those relate:

https://github.com/ppetzold/nestjs-paginate/blob/master/src/paginate.spec.ts#L697
https://github.com/ppetzold/nestjs-paginate/blob/master/src/paginate.spec.ts#L1512

Could you submit a PR with 2 test cases covering your examples?

**Comment by daniel-maegerli** [comment]
Let me know if this is what you expected, here's the PR: https://github.com/ppetzold/nestjs-paginate/pull/539

**Comment by ppetzold** [comment]
is this still an issue with v6+ ?

**Comment by Helveg** [comment]
Yes, the issue persists. The problem arises from left joining the `home` relationship, and then filtering on `home.street IS NULL`. Records that have `home IS NULL` somehow pass the `qb.andWhere('__root_home_rel.street IS NULL')` and are then included because of the left join on `__root_home_rel`.

A workaround for now is to include `where: { home: Not(IsNull()) }` in your pagination config.

I think the best way to fix this is for the `addWhereCondition` in `filter.ts` to include `IS NOT NULL` clauses for all the parent relationships it depends on.

**Comment by guesant** [comment]
hello guys, after an investigation, i found a solution that may work

cc @ppetzold @Helveg 


did someone noticed that because $not is not inside allowedOperations, the parseFilter never returns $not:$null?

https://github.com/ppetzold/nestjs-paginate/blob/df6963117720ce4c2b4bd8bc85feca1746874798/src/filter.ts#L350-L356

## Expected

array that includes the $not and $null combined find operator

![image](https://github.com/user-attachments/assets/43300524-2b95-4e47-af01-94f45eb22843)

## Actual

empty array...

![image](https://github.com/user-attachments/assets/183ca9db-9d58-43db-aa12-1aca1e2ed877)

_by the way i'm using `nestjs-paginate` at version `8.6.3`_

![image](https://github.com/user-attachments/assets/e0b75478-4e10-45f2-ae08-11d25d4a45a8)


the line that ignores the $not is the following:

https://github.com/ppetzold/nestjs-paginate/blob/df6963117720ce4c2b4bd8bc85feca1746874798/src/filter.ts#L277-L279

without that continue when suffix == "$not", it cames to here

![image](https://github.com/user-attachments/assets/f0f43234-db44-43c2-97cf-a8014487ebea)

![image](https://github.com/user-attachments/assets/adfde120-6ac9-4827-a6b6-e71ea2d7e34f)


## so...

when adding a check if to not ignore the `$not` suffix

![image](https://github.com/user-attachments/assets/0cd16a0b-ea02-4460-9eae-4217b5a5b9e9)

it does work properly

![image](https://github.com/user-attachments/assets/1c8075e8-dd0e-47b3-a243-7df7bbe7982d)

![image](https://github.com/user-attachments/assets/fb796f3e-b950-40cb-a73f-fb99df4bd869)

db rows

![image](https://github.com/user-attachments/assets/4709483d-ad52-4114-8f62-beb07c7121d2)

![image](https://github.com/user-attachments/assets/868881e1-9a1f-475d-b215-fda6b2fd17bc)

**Comment by guesant** [comment]
tldr:

https://github.com/ppetzold/nestjs-paginate/blob/df6963117720ce4c2b4bd8bc85feca1746874798/src/filter.ts#L277-L279

```diff
- if (token.suffix && !allowedOperators.includes(token.suffix)) {
+ if (token.suffix && !allowedOperators.includes(token.suffix) && token.suffix !== "$not") {
````

**Comment by guesant** [comment]
@daniel-maegerli, try this workaround:

- add literal `"$not"` into colum's filter array:

```js
filterableColumns: {
  "diarios.id": [FilterOperator.EQ, FilterOperator.NULL, "$not"],
},
```

![image](https://github.com/user-attachments/assets/631d45c2-3640-43f4-accb-6cd816de8170)

**Comment by guesant** [comment]
FilterSuffix.NOT make this works...

this still an issue?

if not, i think it should be better documentend.

**Comment by guesant** [comment]
thats already on the docs, I didn't noticed
